File: leetcode/linkedlist/Linked_List_Cycle/linked_list_cycle.go
package main

import "fmt"

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

// Function to detect if the linked list has a cycle
func hasCycle(head *ListNode) bool {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if slow == fast {
			return true
		}
	}
	return false
}

func main() {
	// Create nodes
	head := &ListNode{Val: 3}
	node2 := &ListNode{Val: 2}
	node0 := &ListNode{Val: 0}
	node4 := &ListNode{Val: -4}

	// Connect nodes to form a cycle: 3 -> 2 -> 0 -> -4 -> 2
	head.Next = node2
	node2.Next = node0
	node0.Next = node4
	node4.Next = node2 // Cycle here

	// Detect cycle
	if hasCycle(head) {
		fmt.Println("Cycle detected")
	} else {
		fmt.Println("No cycle detected")
	}
}

File: leetcode/linkedlist/Linked_List_Cycle/linked_list_cycle.cpp
/**
 * Leeocode 141 | Linked List Cycle
 * Problem: https://leetcode.com/problems/linked-list-cycle/description/
 * Difficulty: Medium
 * Editorial: https://dipanjals-notebook.vercel.app/leetcode/linked-list/linked-list-cycle/
 */

#include <iostream>
#include <cstddef>
using namespace std;

//  Definition for singly-linked list.
 struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
 
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode *slow = head;
        ListNode *fast = head;
        while (fast != NULL && fast->next != NULL) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) {
                return true;
            }
        }
        return false;
    }
};

int main() {
    // Create nodes
    ListNode* head = new ListNode(3);
    ListNode* node2 = new ListNode(2);
    ListNode* node0 = new ListNode(0);
    ListNode* node4 = new ListNode(-4);

    // Connect nodes to form a cycle: 3 -> 2 -> 0 -> -4 -> 2
    head->next = node2;
    node2->next = node0;
    node0->next = node4;
    node4->next = node2; // Cycle here

    // Detect cycle
    Solution solution;
    bool has_cycle = solution.hasCycle(head);

    if (has_cycle) {
        cout << "Cycle detected" << endl;
    } else {
        cout << "No cycle detected" << endl;
    }

    // Clean up memory (Note: in a real-world scenario, you should carefully manage memory, especially with cycles)
    delete head;
    delete node2;
    delete node0;
    delete node4;

    return 0;
}